http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/AwsCrtAsyncHttpStreamAdapter.java

```diff
@@ -0,0 +1,130 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.crt.internal;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.CRT;
+import software.amazon.awssdk.crt.http.HttpClientConnection;
+import software.amazon.awssdk.crt.http.HttpException;
+import software.amazon.awssdk.crt.http.HttpHeader;
+import software.amazon.awssdk.crt.http.HttpRequestBodyStream;
+import software.amazon.awssdk.crt.http.HttpStream;
+import software.amazon.awssdk.crt.http.HttpStreamResponseHandler;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Implements the CrtHttpStreamHandler API and converts CRT callbacks into calls to SDK AsyncExecuteRequest methods
+ */
+@SdkInternalApi
+public class AwsCrtAsyncHttpStreamAdapter implements HttpStreamResponseHandler, HttpRequestBodyStream {
+    private static final Logger log = Logger.loggerFor(AwsCrtAsyncHttpStreamAdapter.class);
+
+    private final HttpClientConnection connection;
+    private final CompletableFuture<Void> responseComplete;
+    private final AsyncExecuteRequest sdkRequest;
+    private final SdkHttpResponse.Builder respBuilder = SdkHttpResponse.builder();
+    private final int windowSize;
+    private final AwsCrtRequestBodySubscriber requestBodySubscriber;
+    private AwsCrtResponseBodyPublisher respBodyPublisher = null;
+
+    public AwsCrtAsyncHttpStreamAdapter(HttpClientConnection connection, CompletableFuture<Void> responseComplete,
+                                        AsyncExecuteRequest sdkRequest, int windowSize) {
+        Validate.notNull(connection, "HttpConnection is null");
+        Validate.notNull(responseComplete, "reqComplete Future is null");
+        Validate.notNull(sdkRequest, "AsyncExecuteRequest Future is null");
+        Validate.isPositive(windowSize, "windowSize is <= 0");
+
+        this.connection = connection;
+        this.responseComplete = responseComplete;
+        this.sdkRequest = sdkRequest;
+        this.windowSize = windowSize;
+        this.requestBodySubscriber = new AwsCrtRequestBodySubscriber(windowSize);
+
+        sdkRequest.requestContentPublisher().subscribe(requestBodySubscriber);
+    }
+
+    private void initRespBodyPublisherIfNeeded(HttpStream stream) {
+        if (respBodyPublisher == null) {
+            respBodyPublisher = new AwsCrtResponseBodyPublisher(connection, stream, responseComplete, windowSize);
+        }
+    }
+
+    @Override
+    public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] nextHeaders) {
```

Comment:
Another question, sometimes services will put `connection:close` in the response header to indicate the connection is to be closed on the service side, would the underlying crt client close such connections after they are released to the pool?

Reply:
The connection will get closed before returning to the pool and the pool will check status and toss it, so it won't be vended again to the next requester

Reply:
I'm going to go ahead and add a closed check on acquire as well, but it may arrive after the fact.

Reply:
Update: this wasn't necessary; the instant a connection enters the closed state, if it's in the pool, it gets removed.

Reply:
To the original question, the crt does not make any interpretations on status code.  Behavioral changes would need to be driven by the SDK.  If we need to make this kind of a change, it will require a little more work, since we don't directly a close on the connection.

Reply:
I see. For sync http clients, the SDK core module has access to the underlying connection, so it is able to close connections that generated a 5xx error. For async http clients, since core module doesn't have access to the connection, individual client would need to close such connections afterwards. This is how we implemented in netty for reference #1768

Reply:
Updated, with test